### DISCUSSION TITLE
!Mode Roadmap!

### DISCUSSION THREAD

**Discussion by IHateMyKite** [IHateMyKite/UnforgivingDevices#38] [Announcements]
This place will work as roadmap for development. This is mainly for me, so I don't add more things than necessary. 

Last update: 16.05.2024

**Latest release = 2.1**
**Next release = 3.0**

### Current plan

- **Version 3.0 will be released after DDNG 0.3 is released, as UD uses native functions from there**

- Total rework of all base scripts. All modules will extend new UD_ModuleBase script. Then, UnforgivingDevicesMain will have array which will store all modules, and work with them (update them, init them, reset them, etc..). This should make further development (adding new modules) simpler. - **Too much effort with little benefit. For now skipping**
- Rework of sentient devices (ability to talk to devices, making kind deals). Addition of new complex modifiers. Will require some kind of monitoring and reward system to be implemented. Small additions to Abadon quest. Level system for devices. - **TODO**
- Rework of modifiers using quest aliases. Then add more modifiers using this new system - **Done**
- Abadon quest (will most likely be part of Unforgiving Souls). But first I will need to do the sentient rework, and modifier rework, as I will need these features for quest - **After Sentient rework. Might do some less complex part of the quest (prologue) which will not use these features yet**
- Rework of orgasm system so it uses native plugin. This will be most complicated part, as It will be impossible to make it compatible with LE. So doing this will mean that mod will no longer support LE. This mod is already heading in this direction, but I'm still hesitating, as big part of users still use LE. Will first do everything else and then decide. I'm also thinking to make this as separate mod which will UD require to work. This way, other mods can also use orgasm system (sexlab for example). But making it as separete mod will require much more effort. - **DONE**

### Future plans
- **3.1** = Dunno yet

### Future plans without date
- More devices with unique color (will probably be in Abadon quest)

**Comment by BreadDain** [comment]
Oh wow, even animations are planned, these are some serious plans. I personally don't know how to make them but have some idea in general, you might also ask some animators on LL like darkconsole, BakaFactory, Billy and etc in case you have some sort of issues with them.

On different device additions: I may suggest this blog https://darkconsole.tumblr.com/post/149294467464/tutorial-making-decent-latex-in-skyrim
I've noticed that rare restraints use only diffuse map replacemers, it is also a good idea to try out different cubic or normal maps to change the desired effect. 
Changing a mesh properties, sadly, is not a good decision since people are using different shapes and morphs for them and new ones will just clip with them, unless these are items that are not affected by it such as gags, plugs, blindfolds and hood.
There is also some unused assets lying around LL such as [Dasha's Devious Toys](https://www.loverslab.com/topic/50289-dashas-devious-toys/) or [KziitdXXXToolset](https://www.loverslab.com/blogs/entry/10533-kziitdxxxtoolset-20100109-fix-a-9-semens-display-error/) which might be worth looking at.

**Comment by kurotatsu77** [comment]
This wasn't updated for a while...

I'm looking for some common agreements and vision of end result.
For example: what are ideas about the mod story? What would happen during it, what result be at end?
Who's Abadon? What is their interest, their character, what is their operation habits?
Mod's devices - are they should remain story quests only? Or we can think and implement some way to manifest them otherwise?
Maybe it's acceptable to some of them, but not for all? Then there should be list of items reserved for quests only.

Current mod is very nice working framework, but without any actual story. It has some fun items (Cursed Goo, for example), but no actual motivation to use them. So my thought that we should try to start developing the story of the mod.

So, as saying goes "criticizing - offer solutions", here's my ideas input, to incite discussion and development:
1. Main quest idea could be PC and Skyrim corruption. The further it goes - the more bondage is spread. This will require to implement some tracking or use existing fame/something else framework.
2. For the quest needed to decide on some activity hub, i.e. place where PC would come again and again, to continue quest
3. At that activity hub should be NPC(s) which offer quests to progress the story etc
4. One possible idea is to use Eorlund Gray-Mane as top-notch blacksmith in search of broadening his skills. PC finds black goo, finds out it's strange effects and searches specialist to consult about it, and as result goes to Eorlund. Then Eorlund with PCs help start to research various ideas about goo use, developing and unlocking all the recipes which are now available from the start. Research could require plenty of goo and gold being supplied, thus motivating PC to get bound in the devices much more. Some research can result in making own base/hub, maybe separate cell under Skyforge.
5. Right now NPC are completely oblivious to PC squirming in restraints and cumming their brains in front of them, which is wrong. Some reactions/actions for NPC should be added. As Skyrim corruption increases, NPC reaction should change from helpful to naughty.

All above is my feeble attempt to provide story quest draft, intended to inspire some ideas, discussion and improvement.

**Reply by IHateMyKite** [reply]
I have some grand addies about the quest but issue is that I should create separate mod for it. UD at this point is already too big, and adding even bigger quest might make the whole project unmaintainable. For that reason, I will create separate mod, which will just named something like UnforgivingDevice - Abadon Quest Continued or something (existing abadon quest will still be present). I don't wan't it to end like DCL, which have so many thing placed in single mod. I might also need to add more dependencies for the Quest (hair mods for NPC, outfits etcc. dunno for now, but adding it to base mod would be stupid).

Regarding the ideas, it looks like we think in the same way. I was planning to first do quest in which player will try to contact "some" person, which might help them find out the origin of the plug/goo/etc.. The person would then send them to receive more goo. After giving the person the goo, they will tell them to return next day (or longer time). After player returns, they will found the person bound in abadon devices. When the player will try to interact with the NPC, it will also lock them in devices. Quest objective would then be to help person and player to escape. I think that this will be some kind of puzzle. Like that player will have a yoke with inescapable locks, and they will have to lockpick locks of the other persons armbinder/straitjacket/whatever. And the quest would continue in similar manner.

After completing the quest, the player will have access buy some random shit from the NPC. 
The NPC will also provide some very simple repeatable quests. I'm mainly thinking about quest in which player will have to find bound npc somewhere (marked on map), and help them (it would not be so simple obviously). That would also supplement your idea of corrupting skyrim.

This is all I'm planning for now. Once this is done, I will focus on next stage.

**Comment by kurotatsu77** [comment]
You're probably right, creating one or several mods/modules is probably better way. Only downside - it means several esps to be in the mods load order, which is limited to 255 and overfilled for some people. But there's a way for them to merge patches, I believe?

Maybe move the Abadon curse to that story mod then, keeping this one as kind of framework?

Well, here's idea for the mod's name - Unforgiving Skyrim.

Do you like idea of research and opening up recipes through it?

Thinking on it, probably the first person PC would contact would be Alchemist in the Whiterun, Aradia, I think. She can be either person which work with us or will point us to our main contact for the story. 

Definitely radiant quests are great addition!

Let's brainstorm together, maybe we will develop something fun!

**Comment by IHateMyKite** [comment]
I would better not move the existing quest, as that would might create issue for some users who already play the mod. For now, I see the current abadon quest more of like build in tutorial. It is very simple and doesn't really affect the framework that much.

The idea with alchemist is great, but Aradia would not work. Better would be to just create new NPC. Using existing npc can create many new issues. Maybe there can be some intermediate quest where player will contact Aradia, then she will point the player to the right person? I will have to also create the NPC location in place that is as compatible as possible. That mostly means that cities are out of questions because there are too much city overwork mods that would break the quest. I will probably use some secret location in wilderness, or even in dungeon (something like necromancer layer, or something). Maybe the alchemist hut close to riften would also work, but there are no doors. Alternatively there is this alchemist hut in blackreach (one with skeleton), which would also work. It would also force player to do some exploring.

And yes, I like the research idea you mentioned. I think that should be very easy to implement. Only issue is that recipes are currently mainly in main mod, so I would have to move them to quest mod.

**Comment by kurotatsu77** [comment]
Another place for the activity hub (and NPC(s)) could be derelict house near destroyed watch tower (where we kill first dragon). We can have one of the projects as to rebuilding it into the proper shop/workshop! 

I think activity hub better to be near Whiterun, like in center of Skyrim, accessible not only by fast travel but also short walk from carriage/stable.

And another point - we actually don't have to move those recipes to the quest mod! We can add overrides into it, if it's installed and enabled - they are available only if certain conditions are met! Same way Abadon plug quest can be overriden to add more options and flavor if the mod is there. Otherwise - it will work as it is now.

Another point to consider is some kind of long lasting effects of losing mind while cumming bound. Maybe simple corruption value being increased by it would be good?

Also, do we have separated behavior of the devices to struggle and using proper keys? I think devices should not activate any of their functions if the proper key used, like for example Abadon plug when unlocked should not attach outfit upon release, various other effects of other devices should not be activated, no taunting of sentient devices etc.

**Comment by IHateMyKite** [comment]
Updated the roadmap. Please let me know what all of you think.

**Reply by kurotatsu77** [reply]
Oh boy. I'm really not thrilled at idea of dropping LE support...

New modifiers sounds interesting. Are shock and bind helper among them?

Abadon plug sure needs some love to it.

**Comment by IHateMyKite** [comment]
Updated after long time the Road map in regards to version 3.0 release

**Reply by kurotatsu77** [reply]
Wow, separate mod for Abadon Quest? I'm really intrigued!
Hopefully the plugs will get more lovely effects?
I had kind of idea about the mind break from it, but never wrote it all, I think.

**Comment by IHateMyKite** [comment]
Planning to release the 3.0 version this week, or on beginning of next week. Let me know if there is any more fatal issues which needs fixing, or if you have any other ideas. It will be released after DDNG 0.4.0 version is also released.

**Reply by kurotatsu77** [reply]
From the top of my head - the issue with patches, they need to be updated. We can't expect users on LL to run Synthesis.
Currently UD doesn't supports inflatable gags (new DD NG ones, which use Dasha's gag mesh) , needs to be added.
Regarding wonderful modifiers rework - will the Synthesis patcher need to be updated?
Also I had some weird behavior in jail with inventory removal, maybe needs some kinds of patch, like you did for MQ107.

**Reply by IHateMyKite** [reply]
> From the top of my head - the issue with patches, they need to be updated. We can't expect users on LL to run Synthesis. Currently UD doesn't supports inflatable gags (new DD NG ones, which use Dasha's gag mesh) , needs to be added. Regarding wonderful modifiers rework - will the Synthesis patcher need to be updated? Also I had some weird behavior in jail with inventory removal, maybe needs some kinds of patch, like you did for MQ107.

Can you please create issues for these issues ? So I dont forget about them. I will update the patches once the 0.4.0 is released, so its most up to date. The new modifier rework does not affect already patched device (but it does affect hand made devices, so you better check US if it does not use some devices with modifiers). All other issues will probably be addressed after 3.0 is released. I dont see them as having big priority.

**Reply by kurotatsu77** [reply]
Created issues #219 and #220.
Don't forget about the xEdit scripts / Synthesis, which are used UDify normal DDs. They most definitely will need update to work with the new modifiers.
Yep, will do my best to bring up US devices to speed.

**Reply by IHateMyKite** [reply]
In the end I removed the patches. DDNG is updating quite often, so I would have to update the patches all the time. That was not the case with old patches as DD was getting update at most once per few years. I have added links to fomod, and will add some tutorial to LL page once its published
